src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/dynamicProgramming/WrappingDPAlgorithm.java

```diff
@@ -0,0 +1,62 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.dynamicProgramming;
+
+import se.liu.ida.hefquin.engine.federation.FederationAccessException;
+import se.liu.ida.hefquin.engine.queryplan.LogicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.LogicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.logical.NaryLogicalOp;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalOperatorForLogicalOperator;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.QueryOptimizationContext;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class WrappingDPAlgorithm {
+    protected final QueryOptimizationContext ctxt;
+
+    public WrappingDPAlgorithm( final QueryOptimizationContext ctxt ) {
+        assert ctxt != null;
+        this.ctxt = ctxt;
+    }
+
+    public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws FederationAccessException {
+        final boolean keepMultiwayJoins = true;
+        final PhysicalPlan initialPhysicalPlan = ctxt.getLogicalToPhysicalPlanConverter().convert(initialPlan, keepMultiwayJoins);
+
+        return optimizePhysicalPlan(initialPhysicalPlan);
+    }
+
+    public PhysicalPlan optimizePhysicalPlan( final PhysicalPlan pp) throws FederationAccessException {
+        final List<PhysicalPlan> children = findChildren(pp);
+        final PhysicalOperatorForLogicalOperator pop = (PhysicalOperatorForLogicalOperator) pp.getRootOperator();
+        final LogicalOperator lop = pop.getLogicalOperator();
+        if ( lop instanceof NaryLogicalOp ) {
```

Comment:
A multiway union is also an NaryLogicalOp but that one should not be passed to the DP algorithm.